File: backend/src/main/kotlin/org/castlekong/backend/entity/Post.kt
package org.castlekong.backend.entity

import jakarta.persistence.CollectionTable
import jakarta.persistence.Column
import jakarta.persistence.ElementCollection
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import java.time.LocalDateTime

@Entity
@Table(name = "posts")
data class Post(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "channel_id", nullable = false)
    val channel: Channel,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    val author: User,
    @Column(nullable = false, columnDefinition = "TEXT")
    val content: String,
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    val type: PostType = PostType.GENERAL,
    @Column(name = "is_pinned", nullable = false)
    val isPinned: Boolean = false,
    @Column(name = "view_count", nullable = false)
    val viewCount: Long = 0,
    @Column(name = "like_count", nullable = false)
    val likeCount: Long = 0,
    @Column(name = "comment_count", nullable = false)
    val commentCount: Long = 0,
    @Column(name = "last_commented_at")
    val lastCommentedAt: LocalDateTime? = null,
    @ElementCollection(targetClass = String::class, fetch = FetchType.LAZY)
    @CollectionTable(name = "post_attachments", joinColumns = [JoinColumn(name = "post_id")])
    @Column(name = "file_url", nullable = false, length = 500)
    val attachments: Set<String> = emptySet(),
    @Column(name = "created_at", nullable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),
    @Column(name = "updated_at", nullable = true)
    val updatedAt: LocalDateTime? = null,
)

enum class PostType {
    GENERAL,
    ANNOUNCEMENT,
    QUESTION,
    POLL,
}

File: frontend/lib/features/post/domain/entities/post.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'author.dart';

part 'post.freezed.dart';
part 'post.g.dart';

/// 게시글을 나타내는 불변 Entity
///
/// Slack 스타일로 제목 없이 내용만 포함합니다.
@freezed
class Post with _$Post {
  const factory Post({
    /// 게시글 고유 ID
    required int id,

    /// 게시글 내용 (제목 없음)
    required String content,

    /// 작성자 정보
    required Author author,

    /// 작성 시각
    required DateTime createdAt,

    /// 수정 시각 (수정되지 않았으면 null)
    DateTime? updatedAt,

    /// 댓글 수
    @Default(0) int commentCount,

    /// 마지막 댓글 작성 시각 (댓글이 없으면 null)
    DateTime? lastCommentedAt,
  }) = _Post;

  /// JSON에서 Post 객체 생성
  factory Post.fromJson(Map<String, dynamic> json) => _$PostFromJson(json);
}
